lec03-basic-r.Rmd

```diff
@@ -1,4 +1,732 @@
+```
+
+Integer refers to a whole number, such as 1, 2, 3, 4, etc. Numbers with
+decimals, 1.0, 2.4, 3.333, are referred to as floats. Factors are used to
+represent categorical data. Factors can be ordered or unordered, and
+understanding them is necessary for statistical analysis and for plotting.
+Factors are stored as integers, and have labels (text) associated with these
+unique integers. While factors look (and often behave) like character vectors,
+they are actually integers under the hood, and you need to be careful when
+treating them like strings.
+
+
+## Inspecting `data.frame` Objects
+
+We already saw how the functions `head()` and `str()` can be useful to check the
+content and the structure of a data frame. Here is a non-exhaustive list of
+functions to get a sense of the content/structure of the data. Let's try them
+out!
+
+* Size:
+    * `dim(surveys)` - returns a vector with the number of rows in the first element,
+          and the number of columns as the second element (the **dim**ensions of
+          the object)
+    * `nrow(surveys)` - returns the number of rows
+    * `ncol(surveys)` - returns the number of columns
+
+* Content:
+    * `head(surveys)` - shows the first 6 rows
+    * `tail(surveys)` - shows the last 6 rows
+
+* Names:
+    * `names(surveys)` - returns the column names (synonym of `colnames()` for `data.frame`
+	   objects)
+    * `rownames(surveys)` - returns the row names
+
+* Summary:
+    * `str(surveys)` - structure of the object and information about the class, length and
+	   content of  each column
+    * `summary(surveys)` - summary statistics for each column
+
+Note: most of these functions are "generic", they can be used on other types of
+objects besides `data.frame`.
+
+> ### Challenge
+>
+> Based on the output of `str(surveys)`, can you answer the following questions?
+>
+> * What is the class of the object `surveys`?
+> * How many rows and how many columns are in this object?
+> * How many species have been recorded during these surveys?
+
+<!---
+```{r, echo=FALSE}
+## Answers
+##
+## * class: data frame
+## * how many rows: 34786,  how many columns: 13
+## * how many species: 48
+```
+--->
+
+
+## Indexing and subsetting data frames
+
+Our survey data frame has rows and columns (it has 2 dimensions), if we want to
+extract some specific data from it, we need to specify the "coordinates" we want
+from it. Row numbers come first, followed by column numbers. However, note that
+different ways of specifying these coordinates lead to results with different
+classes.
+
+```{r}
+surveys[1, 1]   # first element in the first column of the data frame (as a vector)
+surveys[1, 6]   # first element in the 6th column (as a vector)
+surveys[, 1]    # first column in the data frame (as a vector)
+surveys[1]      # first column in the data frame (as a data.frame)
+surveys[1:3, 7] # first three elements in the 7th column (as a vector)
+surveys[3, ]    # the 3rd element for all columns (as a data.frame)
```

Comment:
You have to be careful here. All of these indexing return a `data.frame`/`tbl_df`, **not** a vector. That's because this is a `tibble` object and so will always return a `tibble` object *unless* you use `[[` indexing. The reason is because the tibble package is trying to remove some strange inconsistencies (as you describe).

Reply:
Ah, nice catch. I originally was working with standard data frames durign the first part of the lecture. When I switch to introducing tibbles right from the beginning, I forgot that it would impact this indexing since it consistently always returns a tibble as you said. Will mention this convenient consistency and remove the erroneous comments.